File: app/tools/magic_wand_tool.py
import numpy as np
import cv2
from PySide6.QtCore import QPointF, Qt
from PySide6.QtGui import QMouseEvent, QPainter, QPen, QColor
from app.tools.base_tool import BaseTool

class MagicWandTool(BaseTool):
    """
    Smart Color Selection / Magic Wand tool.
    Clicking a location flood-fills pixels with similar color within tolerance
    and updates the alpha mask (erasing background).
    """

    def __init__(self, canvas):
        super().__init__(canvas)
        self.tolerance = 30        # 1 to 100
        self.contiguous = True     # Flood fill vs global color selection
        self.feather = 2

    def mouse_press(self, img_pos: QPointF, event: QMouseEvent):
        if event.button() != Qt.MouseButton.LeftButton or not self.canvas.document:
            return

        doc = self.canvas.document
        lyr = doc.active_layer
        if not lyr or lyr.image is None:
            return

        img = lyr.image
        h, w = img.shape[:2]

        layer_pos = doc.map_canvas_pos_to_layer_pos(img_pos, lyr)
        seed_x, seed_y = int(round(layer_pos[0])), int(round(layer_pos[1]))

        if seed_x < 0 or seed_x >= w or seed_y < 0 or seed_y >= h:
            return

        # Prepare new mask for active layer
        mask = lyr.mask.copy() if lyr.mask is not None else np.full((h, w), 255, dtype=np.uint8)

        if self.contiguous:
            # OpenCV Flood Fill
            ff_mask = np.zeros((h + 2, w + 2), dtype=np.uint8)
            flags = 4 | (255 << 8) | cv2.FLOODFILL_MASK_ONLY | cv2.FLOODFILL_FIXED_RANGE
            lo_diff = (self.tolerance, self.tolerance, self.tolerance)
            up_diff = (self.tolerance, self.tolerance, self.tolerance)

            cv2.floodFill(img, ff_mask, (seed_x, seed_y), 0, lo_diff, up_diff, flags)
            selected_region = ff_mask[1:h+1, 1:w+1] > 0
        else:
            # Global color distance in LAB space
            seed_color = img[seed_y, seed_x].astype(np.float32)
            dist = np.linalg.norm(img.astype(np.float32) - seed_color, axis=2)
            selected_region = dist <= self.tolerance

        # Erase selected region from layer mask
        mask[selected_region] = 0

        # Push command
        doc.update_mask(mask, layer_id=lyr.id, description="Magic Wand Select")
        self.canvas.update()


    def mouse_move(self, img_pos: QPointF, event: QMouseEvent):
        pass

    def mouse_release(self, img_pos: QPointF, event: QMouseEvent):
        pass

    def draw_overlay(self, painter: QPainter):
        """Draws small target crosshair cursor for magic wand."""
        if self.canvas.hover_img_pos is not None:
            pos = self.canvas.hover_img_pos
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)
            pen = QPen(QColor(255, 255, 0), 1.5, Qt.PenStyle.DashLine)
            painter.setPen(pen)
            painter.drawEllipse(pos, 8, 8)
            painter.drawLine(int(pos.x() - 12), int(pos.y()), int(pos.x() + 12), int(pos.y()))
            painter.drawLine(int(pos.x()), int(pos.y() - 12), int(pos.x()), int(pos.y() + 12))

File: Cpp/src/tools/MagicWandTool.cpp
#include "tools/MagicWandTool.h"
#include "ui/CanvasView.h"
#include "core/ImageDocument.h"
#include "core/Layer.h"
#include <opencv2/imgproc.hpp>
#include <QPen>
#include <QColor>
#include <cmath>

namespace ImageCut {
namespace Tools {

MagicWandTool::MagicWandTool(UI::CanvasView* canvas) : BaseTool(canvas) {}

void MagicWandTool::mousePress(const QPointF& imgPos, QMouseEvent* event) {
    if (event->button() != Qt::LeftButton || !m_canvas || !m_canvas->getDocument()) return;

    auto doc = m_canvas->getDocument();
    auto lyr = doc->getActiveLayer();
    if (!lyr || lyr->image.empty()) return;

    cv::Mat img = lyr->image;
    int h = img.rows;
    int w = img.cols;

    auto [lx, ly] = doc->mapCanvasPosToLayerPos(imgPos, lyr);
    int seedX = static_cast<int>(std::round(lx));
    int seedY = static_cast<int>(std::round(ly));

    if (seedX < 0 || seedX >= w || seedY < 0 || seedY >= h) return;

    cv::Mat mask = !lyr->mask.empty() ? lyr->mask.clone() : cv::Mat(h, w, CV_8UC1, cv::Scalar(255));
    cv::Mat selectedRegion(h, w, CV_8UC1, cv::Scalar(0));

    if (m_contiguous) {
        cv::Mat ffMask(h + 2, w + 2, CV_8UC1, cv::Scalar(0));
        int flags = 4 | (255 << 8) | cv::FLOODFILL_MASK_ONLY | cv::FLOODFILL_FIXED_RANGE;
        cv::Scalar diff(m_tolerance, m_tolerance, m_tolerance);

        cv::Mat rgbImg;
        if (img.channels() == 4) {
            cv::cvtColor(img, rgbImg, cv::COLOR_RGBA2RGB);
        } else {
            rgbImg = img;
        }

        cv::floodFill(rgbImg, ffMask, cv::Point(seedX, seedY), cv::Scalar(0), nullptr, diff, diff, flags);
        selectedRegion = ffMask(cv::Rect(1, 1, w, h)).clone();
    } else {
        cv::Mat rgbImg;
        if (img.channels() == 4) {
            cv::cvtColor(img, rgbImg, cv::COLOR_RGBA2RGB);
        } else {
            rgbImg = img;
        }

        cv::Vec3b seedColor = rgbImg.at<cv::Vec3b>(seedY, seedX);
        for (int y = 0; y < h; ++y) {
            for (int x = 0; x < w; ++x) {
                cv::Vec3b c = rgbImg.at<cv::Vec3b>(y, x);
                double dist = std::sqrt(
                    (c[0] - seedColor[0]) * (c[0] - seedColor[0]) +
                    (c[1] - seedColor[1]) * (c[1] - seedColor[1]) +
                    (c[2] - seedColor[2]) * (c[2] - seedColor[2])
                );
                if (dist <= m_tolerance) {
                    selectedRegion.at<uint8_t>(y, x) = 255;
                }
            }
        }
    }

    for (int y = 0; y < h; ++y) {
        for (int x = 0; x < w; ++x) {
            if (selectedRegion.at<uint8_t>(y, x) > 0) {
                mask.at<uint8_t>(y, x) = 0;
            }
        }
    }

    doc->updateMask(mask, lyr->id, "Magic Wand Select");
    m_canvas->updateViewport();
}

void MagicWandTool::mouseMove(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos; (void)event;
}

void MagicWandTool::mouseRelease(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos; (void)event;
}

void MagicWandTool::drawOverlay(QPainter* painter) {
    if (!m_canvas || !painter) return;
    QPointF pos = m_canvas->getHoverPos();
    if (pos.x() >= 0 && pos.y() >= 0) {
        painter->setRenderHint(QPainter::Antialiasing);
        QPen pen(QColor(255, 255, 0), 1.5, Qt::DashLine);
        painter->setPen(pen);
        painter->drawEllipse(pos, 8, 8);
        painter->drawLine(QPointF(pos.x() - 12, pos.y()), QPointF(pos.x() + 12, pos.y()));
        painter->drawLine(QPointF(pos.x(), pos.y() - 12), QPointF(pos.x(), pos.y() + 12));
    }
}

} // namespace Tools
} // namespace ImageCut
